Support for external authorization grpc service.

saumoh opened the pull request:
_Title_
Authorization TCP and HTTP filters using an external gRPC service.
Patch 1 of 3: gRPC client and common framework.

_Description_
As per the feedback on https://github.com/envoyproxy/envoy/pull/2359 I am breaking it up into three PR's.

The patch in this PR adds support for gRPC client that will be used by TCP and HTTP filters to make authorization checks with an external service. 
This PR implements the discussing we had in https://github.com/envoyproxy/envoy/issues/2291.

_Testing_
I have tested this on my local setup with an external gRPC authorization server. I have also added UT's for the gRPC client side.

_Risk_
Low: Because only the users of this filter will be impacted. It should not impact general stability of Envoy it self.

_Caveats_
* I did not test with a Google gRPC service configuration.
* I am aware that https://github.com/envoyproxy/envoy/pull/2393 will impact this PR. After that PR is merged into master I'll merge this branch with master and make the required changes to the gRPC client implementation.

Signed-off-by: Saurabh Mohan <saurabh+github@tigera.io>

htuch commented:
Thanks @saumoh for the PR series. Now that #2393 is merged, can you rebase this and use the client factory for gRPC services provided there?

louiscryan commented:
@qiwzhang can you follow along here so that we are confident about
convergence with https://github.com/istio/proxy/tree/master/src/envoy/mixer

On Tue, Jan 23, 2018 at 9:02 AM, htuch <notifications@github.com> wrote:

> Thanks @saumoh <https://github.com/saumoh> for the PR series. Now that
> #2393 <https://github.com/envoyproxy/envoy/pull/2393> is merged, can you
> rebase this and use the client factory for gRPC services provided there?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/envoyproxy/envoy/pull/2415#issuecomment-359858874>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AIoKPD6enrLseQ8Evbk1u2Ol5NuPRxp3ks5tNhCcgaJpZM4RkxoI>
> .
>

saumoh commented:
@htuch I've merged with the new gRPC manager changes.

htuch commented:
@saumoh can you get the build to pass? I'll take an in-depth pass tomorrow, but I think it's helpful to see the CI build passing. Thanks.

htuch commented on source/common/ext_authz/ext_authz_impl.cc:
```diff
@@ -0,0 +1,242 @@
+  headers.iterate(fillHttpHeaders, mhdrs);
+
+  std::unique_ptr<AttributeContext_Request> req = std::unique_ptr<AttributeContext_Request>{new AttributeContext_Request()};
+  ASSERT(req);
```
Very brief drive-by nits here:
1. Don't need to `ASSERT` on a basic memory allocation, we assume memory allocations always succeed or the world ends in Envoy.
2. Should write this as `auto req = std::make_unique<Attribute_Request>();` for C++14.

htuch commented on source/common/ext_authz/ext_authz_impl.h:
```diff
@@ -0,0 +1,121 @@
+class ExtAuthzCheckRequestGenerator: public CheckRequestGenerator {
+public:
+  ExtAuthzCheckRequestGenerator() {}
+  ~ExtAuthzCheckRequestGenerator() {}
```
Quick nit: the default constructor/destructors are provided by the compiler, no need to add them.

saumoh commented:
@htuch Fixed the build and addressed your last set of comments. If you could please have a look. I am working on fixing coverage in the mean while. thanks.

htuch commented:
@saumoh there is coverage regression here, see `Code coverage 97.7 is lower than limit of 98.0` in CI. Have you generated a full coverage report and looked to see where you are missing tests?

htuch commented on bazel/repositories.bzl:
```diff
@@ -197,6 +199,14 @@ def _envoy_api_deps():
+    auth_bind_targets = [
+        "auth"
+    ]
+    for t in auth_bind_targets:
```
FWIW, the recent merges of https://github.com/envoyproxy/data-plane-api/pull/421 and https://github.com/envoyproxy/envoy/pull/2430 will break this (and other parts of your PR). Be prepared to update when you merge master or pull in an updated data-plane-api SHA.

saumoh replied:
Thanks for the heads up. I'll do another merge with master to pull in the latest data-plane-api and rework the impacted code.

htuch commented on include/envoy/ext_authz/ext_authz.h:
```diff
@@ -0,0 +1,98 @@
+  /**
+   * Called when a check request is complete. The resulting status is supplied.
+   */
+  virtual void complete(CheckStatus status) PURE;
```
Nit: I think `onComplete` is a clearer name.

htuch commented on include/envoy/ext_authz/ext_authz.h:
```diff
@@ -0,0 +1,98 @@
+  virtual ~ClientFactory() {}
+
+  /**
+   * Return a new authz client.
```
@param docs for all parameters on `include/envoy` interface headers.

htuch replied:
Also @return for the return type & text.

htuch commented on include/envoy/ext_authz/ext_authz.h:
```diff
@@ -0,0 +1,98 @@
+ * and fill out the details in the authorization protobuf that is sent to authorization
+ * service.
+ */
+class CheckRequestGenerator {
```
This is really a helper utility class. Generally we don't put these in `include/envoy`. A litmus test is "will I need to mock this class?" and "will this class have potentially more than one implementation?". If no to both, put it in the `source/common` subtree.

saumoh replied:
Initially I had it in the `souce/common` subtree. When I started writing the UT's for the filters I had to [mock](https://github.com/colabsaumoh/envoy/blob/141a203e47df1ba097e056b9395d8d47a0864f3b/test/mocks/ext_authz/mocks.h#L25) it. That's when I moved it here.
The usage in http filter is [here](
https://github.com/colabsaumoh/envoy/blob/fe93fd61b6ebb73247b089f47c7ea48d75e6af73/test/common/http/filter/ext_authz_test.cc#L117)

I think that the mock is needed. @htuch Should I just keep it here or move to `source/common` and have the mock reference it from there? please confirm. thx.

htuch replied:
Do you really need to mock the proto creation? These are basically straight line pure functions. Personally, I would mock the input parameters, since it's just as easy to do and results in easier to understand code.

saumoh replied:
i think i know what you mean here. Let me try to redo the ut's without using the mock.

saumoh replied:
@htuch last commit removes the mock for proto creation.

htuch commented on source/common/access_log/grpc_access_log_impl.cc:
```diff
@@ -131,6 +131,12 @@ void HttpGrpcAccessLog::responseFlagsToAccessLogResponseFlags(
+
+  // TODO(saumoh): Uncomment once unauthorization has been added to accesslog.proto
+  //
+  // if (request_info.getResponseFlag(RequestInfo::ResponseFlag::Unauthorized)) {
```
I would remove all of this for now and keep it locally. It will bit rot otherwise.

htuch commented on source/common/ext_authz/ext_authz_impl.cc:
```diff
@@ -0,0 +1,241 @@
+  async_client_factory_ = async_client_manager.factoryForGrpcService(grpc_service, scope);
+}
+
+ClientPtr GrpcFactoryImpl::create(const Optional<std::chrono::milliseconds>& timeout) {
```
I don't think you need the `GrpcFactoryImpl` if all you're doing is wrapping the `Grpc::AsyncClientManager/Factory` calls, just use those classes directly.

saumoh commented:
@htuch I am adding more tests for some of the newly added code. That should help creep the numbers closer to 100%.

htuch commented on source/common/ext_authz/ext_authz_impl.cc:
```diff
@@ -0,0 +1,241 @@
+std::unique_ptr<::envoy::api::v2::Address> ExtAuthzCheckRequestGenerator::getProtobufAddress(
+    const Network::Address::InstanceConstSharedPtr& instance) {
+
+  auto addr = std::make_unique<::envoy::api::v2::Address>();
```
I think it's more natural in protobuf to invert this construction. What you have is: 
1. getConnectionPeer calls getProtobufAddress to get some remote address PB.
2. getConnectionPeer then does a set_allocated_address with the result.

A much more common pattern is.
1. getConnectionPeer calls `mutable_allocated_addresss()` to allocate memory and obtain a pointer.
2. getProtobufAddress takes this pointer and then writes directly into the memory.

Of course, you would probably rename the methods. See https://github.com/envoyproxy/envoy/blob/master/source/common/config/address_json.cc#L10 and its callers for an example.

What you have is correct, I just find it noisier, since we have to deal with smart pointer creation and release, rather than idiomatically having the protobuf library `mutable_foo()` allocators do their work.

htuch commented on source/common/ext_authz/ext_authz_impl.cc:
```diff
@@ -0,0 +1,241 @@
+  req->set_allocated_http(httpreq);
+
+  return req;
+}
```
I think it would be instructive to compare with https://github.com/envoyproxy/envoy/blob/master/source/common/access_log/grpc_access_log_impl.cc#L251. There is a lot of overlap between the access log proto data and construction and the auth request.

As a non-actionable reflection, it would have been great if we could have converged the common objects for both access logs and auth, but I don't think this is going to happen (@mattklein123).

wora replied:
@htuch This really depends on how big the scale we are targeting. If we target millions of services, billions of requests per second, we can surely refactor the code, but it will take quite a bit work.

The general approach would be something like this:
- Reach an agreement on the standard Envoy attributes, and define them as a proto like AttributeContext.
- Envoy natively generate the standard attributes.
- Design the standard logs, metrics, events, traces , and other telemetry schemas.
- Generate standard telemetry data based on the standard attributes.
- Optional. Generate legacy telemetry data, such as access logs, from the standard attributes.

New users of Envoy won't need the legacy telemetry data, so it would just be disabled by a flag. We can delete the legacy code after a year, or just keep them on the side.

This refactoring would only be useful if we can agree on a rich set of standard telemetry data that appeal to wide range of users, so a customer can correlate telemetry data across services written by different vendors.

saumoh replied:
@htuch Ref to the first paragraph, are you suggesting that I write this in a way that it's done in https://github.com/envoyproxy/envoy/blob/master/source/common/access_log/grpc_access_log_impl.cc#L251.
I am happy to do that but just want to confirm.

Separately, thanks for all the other review comments. I am working on making those changes.

htuch replied:
@wora agreed, I'm just waxing philosophical. 

@saumoh just suggesting you give that code a pass and see if there's anything in there that your code would benefit from following.

htuch commented on source/common/ext_authz/ext_authz_impl.h:
```diff
@@ -0,0 +1,116 @@
+  ClientPtr create(const Optional<std::chrono::milliseconds>&) override {
+    return ClientPtr{new NullClientImpl()};
+  }
+};
```
I'm guessing you modeled this on the RateLimiterImpl, do you really need a null implementation?

saumoh commented:
@htuch Thanks for working with me on this PR. I've updated with the last set of pr comments.
* Merged with latest data-plane-api re-org  https://github.com/envoyproxy/data-plane-api/pull/421
* Moved the `CheckRequestGenerator` out of the `include/` to `source/common`
* Added more UT's
* Rewrote the proto buf creation to avoid creating/destroying smart pointers (thanks for the reference code)
* Addressed rest of the comments from the previous reviews.

htuch commented on include/envoy/ext_authz/ext_authz.h:
```diff
@@ -0,0 +1,67 @@
+  /**
+   * Cancel an inflight Check request.
+   */
+  virtual void cancel() PURE;
```
Question: can there only ever be a single inflight request per client with this interface? Does this imply that each request must have its own `Client`?

saumoh replied:
I was thinking that the per-thread gRPC would save us setup time and hence would help even in the unary gRPC case. But if the cluster is going to do that for us then I agree changing it would not buy us much.

htuch replied:
Yeah, for Envoy gRPC, this will come for free from the cluster pool management. In theory, Google gRPC also does this, based on good authority, but I haven't validated. Let's work on your other PRs first and we can circle back to this later.

htuch commented on source/common/ext_authz/ext_authz_impl.cc:
```diff
@@ -0,0 +1,218 @@
+}
+
+void ExtAuthzCheckRequestGenerator::addressToProtobufAddress(
+    envoy::api::v2::Address& proto_address, const Network::Address::Instance& address) {
```
Tiny nit: prefer input params first, output params following.

htuch commented on source/common/ext_authz/ext_authz_impl.cc:
```diff
@@ -0,0 +1,218 @@
+  callbacks_ = nullptr;
+}
+
+void ExtAuthzCheckRequestGenerator::addressToProtobufAddress(
```
Should this be a feature of `Network::Address::Instance` or in https://github.com/envoyproxy/envoy/blob/master/source/common/network/utility.h? Seems generally useful.

htuch commented on source/common/ext_authz/ext_authz_impl.cc:
```diff
@@ -0,0 +1,218 @@
+const std::string ExtAuthzCheckRequestGenerator::getProtocolStr(const Envoy::Http::Protocol& p) {
+
+  switch (p) {
+  case Envoy::Http::Protocol::Http10:
```
This utility belong probably somewhere with `Envoy::Http::Protocol`.

htuch commented on source/common/ext_authz/ext_authz_impl.cc:
```diff
@@ -0,0 +1,218 @@
+  }
+}
+
+const std::string ExtAuthzCheckRequestGenerator::getProtocolStr(const Envoy::Http::Protocol& p) {
```
`const` qualifier here is not needed (it doesn't strengthen the type). Also, prefer `getProtocolString` as a nit, we try and avoid abbreviations where things don't get massively verbose.

htuch commented on source/common/ext_authz/ext_authz_impl.cc:
```diff
@@ -0,0 +1,218 @@
+}
+
+Envoy::Http::HeaderMap::Iterate
+ExtAuthzCheckRequestGenerator::fillHttpHeaders(const Envoy::Http::HeaderEntry& e, void* ctx) {
```
This can just be a lamda inlined with the `iterate` that uses it`

htuch commented on source/common/ext_authz/ext_authz_impl.cc:
```diff
@@ -0,0 +1,218 @@
+
+Envoy::Http::HeaderMap::Iterate
+ExtAuthzCheckRequestGenerator::fillHttpHeaders(const Envoy::Http::HeaderEntry& e, void* ctx) {
+  Envoy::Protobuf::Map<::std::string, ::std::string>* mhdrs =
```
Nit: prefer `mutable_headers`

htuch commented on source/common/ext_authz/ext_authz_impl.cc:
```diff
@@ -0,0 +1,218 @@
+ExtAuthzCheckRequestGenerator::fillHttpHeaders(const Envoy::Http::HeaderEntry& e, void* ctx) {
+  Envoy::Protobuf::Map<::std::string, ::std::string>* mhdrs =
+      static_cast<Envoy::Protobuf::Map<::std::string, ::std::string>*>(ctx);
+  (*mhdrs)[std::string(e.key().c_str(), e.key().size())] =
```
Maybe we should just add to `HeaderString` etc. a `std::string` accessor?

htuch commented on source/common/ext_authz/ext_authz_impl.cc:
```diff
@@ -0,0 +1,218 @@
+
+std::string ExtAuthzCheckRequestGenerator::getHeaderStr(const Envoy::Http::HeaderEntry* entry) {
+  if (entry) {
+    return std::string(entry->value().c_str(), entry->value().size());
```
Ditto

htuch commented on source/common/ext_authz/ext_authz_impl.h:
```diff
@@ -0,0 +1,113 @@
+typedef ConstSingleton<ConstantValues> Constants;
+
+// TODO(htuch): We should have only one client per thread, but today we create one per filter stack.
+// This will require support for more than one outstanding request per client.
```
Yeah, I see you're following the `RateLimit` pattern here. I'm not sure if it's the ideal one to copy, given that it has these limits. Probably fine for MVP, but it's not great that we're building on this.

saumoh replied:
@hutch is there reference where there is a client per thread in existing code. I followed the `RateLimit` since it seemed closest to what `ExtAuthz` needed to do.
If it does not exist and if the patter of invoking external services is becoming common-which i suspect it is- should I open a issue/enhancement to track creating a gRPC client per thread?

htuch replied:
You can look at what https://github.com/envoyproxy/envoy/blob/master/source/common/access_log/grpc_access_log_impl.h does, I think that's close to what you want.

saumoh replied:
thanks for the reference. let me take a look at it.

saumoh replied:
@htuch I have a quick question on this. afaik the gRPC request is per thread and multiple `listener` filter stacks will use it. Will gRPC stack be able to take care of co-relating the `response` with the correct `request`? Or is this something that the `ThreadLocalCallers` class will have to handle with a `unordered_map` like [here]( (https://github.com/envoyproxy/envoy/blob/5f8b84179ffc1530476da4d8a250835fa1274a3c/source/common/access_log/grpc_access_log_impl.h#L101)
Even with that there is not a unique identifier in the gRPC request/response for ExtAuthz. So it's not clear to me how to map the response back to the calling/invoker filter.
Unless if gRPC already takes care of this under the covers?

htuch replied:
As mentioned above, I think it's different for authz, since you are unary RPC, so you can probably put this aside for now.

htuch commented on source/common/ext_authz/ext_authz_impl.h:
```diff
@@ -0,0 +1,113 @@
+   * @param request is the reference to the check request that will be filled up.
+   *
+   */
+  void createHttpCheck(const Envoy::Http::StreamDecoderFilterCallbacks* callbacks,
```
Should these all be `static`? It's not so much a generator as a collection of static utility functions.

wora commented on source/common/ext_authz/ext_authz_impl.cc:
```diff
@@ -0,0 +1,218 @@
+  case Envoy::Http::Protocol::Http11:
+    return std::string("Http1.1");
+  case Envoy::Http::Protocol::Http2:
+    return std::string("Http2");
```
These strings should be documented in the attribute context proto.

wora commented on source/common/ext_authz/ext_authz_impl.cc:
```diff
@@ -0,0 +1,218 @@
+  default:
+    break;
+  }
+  return std::string("unknown");
```
We should use native proto3 semantics for unknown, e.g. all unknown should be represented as 0, false, "".

For example, I need to query logs for unauthenticated users, I naturally write `user == ""` instead of `user == "unknown"`, the same applies to protocol, ip address, user agent. It is better to use native proto3 default to represent all unknowns.

saumoh commented:
@htuch. @wora Thanks for the review comments. How do you feel about doing the generalizations such as separating out the `attribute_context` creation into a separate/different PR? There is a bunch of commonality we've already observed like [here](https://github.com/envoyproxy/envoy/blob/4548c6520c730cacd84ac00023b5bbf466fa1c05/source/common/access_log/grpc_access_log_impl.cc#L67) and [here](https://github.com/envoyproxy/envoy/blob/4548c6520c730cacd84ac00023b5bbf466fa1c05/source/common/access_log/grpc_access_log_impl.cc#L196) that i think should be handled in a new/different PR.
I would be happy to work on those PR's. To me some of this work feels out of scope of this specific PR.
wdyt?

htuch commented:
@saumoh sure, they can be separate PRs. Do you want to do them first? They should be very small and quick to get through review.

saumoh commented:
@htuch i should have been clearer in what i meant. I was indicating that I will make the changes that can use the new common code in another pr.

htuch commented:
I think if you can do a few small PRs for extra utilities for header strings and network address proto that would be good to do now; the more general separation of attribute context stuff I'll leave for you and @wora to discuss.

wora commented on source/common/ext_authz/ext_authz_impl.cc:
```diff
@@ -0,0 +1,189 @@
+  }
+
+  // Fill in the headers
+  auto mhdrs = httpreq.mutable_headers();
```
We need a way to control which headers to send.

wora commented on source/common/ext_authz/ext_authz_impl.cc:
```diff
@@ -0,0 +1,189 @@
+      }
+    }
+  }
+  peer.set_principal(principal);
```
We probably need some format normalization here. At least document what is the format used here?

wora commented on source/common/ext_authz/ext_authz_impl.cc:
```diff
@@ -0,0 +1,189 @@
+  peer.set_principal(principal);
+
+  if (!service.empty()) {
+    peer.set_service(service);
```
Also document the format?

saumoh replied:
@wora: The fulfillment of the `AttributeContext` is as-per the documentation in the data-plane-api specification [1](https://github.com/envoyproxy/data-plane-api/blob/a07a36853e698d5fee7b011d7abd1dcb87eda45e/envoy/service/auth/v2/attribute_context.proto#L40). The code here is adhering to that interface agreement.
I can make a note of that at the top of this class.

saumoh commented:
@htuch added a new pr https://github.com/envoyproxy/envoy/pull/2504 for the changes for address and http changes.

saumoh commented:
@htuch @wora For the separation of the `AttributeContext` in the data-plane-api [1](https://github.com/envoyproxy/data-plane-api/blob/master/envoy/service/auth/v2/attribute_context.proto)
Should I create a utility.[cc|h] in a **new** directory `source/common/auth` which will have a static class `CreateProtoAttributeContext` that will fill up the protobuf.
How do u suggest this be structured in the code? Thanks.

htuch commented:
I'm not sure if we need to refactor it today; it's done at the API level already, we don't have any extra consumers yet.

@wora if you feel strongly about pulling the AttributeContext code out of this PR, can you enumerate some other potential users in the Envoy code base? That will allow us to see how this is going to be used.

We know it's not going to be used for logging, so I'm guessing it's for other auth filters?

htuch commented:
Please merge master following https://github.com/envoyproxy/envoy/pull/2495. @saumoh is this ready for another review pass?

saumoh commented:
@htuch Will merge #2495. thanks for the heads up.
The current branch has all the changes except for per-thread grpc client. I am working on a prototype for it. Do you think I should do the per-thread grpc client post-merge of this PR? I feel that would be a big enough change that it merits a more focused pr. wdyt?

htuch commented:
@saumoh Yeah, you can do that as a followup PR (I did something similar with the Google gRPC client library PRs, it makes it easier to separate out reasoning about thread stuff from the semantics of the changes).

saumoh commented:
@htuch I merged with #2495 and think i've addressed all the review comments.

htuch commented on source/common/ext_authz/ext_authz_impl.cc:
```diff
@@ -0,0 +1,191 @@
+  Envoy::Http::StreamDecoderFilterCallbacks* cb =
+      const_cast<Envoy::Http::StreamDecoderFilterCallbacks*>(callbacks);
+
+  std::string service = getHeaderStr(headers.EnvoyDownstreamServiceCluster());
```
Nit: `const std::string`

htuch commented on source/common/ext_authz/ext_authz_impl.h:
```diff
@@ -0,0 +1,112 @@
+   * createHttpCheck is used to extract the attributes from the stream and the http headers
+   * and fill them up in the CheckRequest proto message.
+   * @param callbacks supplies the Http stream context from which data can be extracted.
+   * @param headers supplies the heeader map with http headers that will be used to create the
```
Nit: s/heeader/header/

htuch commented on source/common/ext_authz/ext_authz_impl.h:
```diff
@@ -0,0 +1,112 @@
+  /**
+   * createTcpCheck is used to extract the attributes from the network layer and fill them up
+   * in the CheckRequest proto message.
+   * @param callbacks supplies the networklayer context from which data can be extracted.
```
Nit: s/networklayer/network layer/

htuch commented on test/common/ext_authz/ext_authz_impl_test.cc:
```diff
@@ -0,0 +1,193 @@
+    client_.onSuccess(std::move(response), span_);
+  }
+
+  {
```
Can you split this into multiple tests? The scoped-grouping-of-a-bunch-of-related-tests is fine for small simple tests, but I think here it would be clearer to have them as distinct tests.

htuch commented on test/common/ext_authz/ext_authz_impl_test.cc:
```diff
@@ -0,0 +1,193 @@
+
+    client_.onCreateInitialMetadata(headers);
+
+    response.reset(new envoy::service::auth::v2::CheckResponse());
```
Nit: prefer `std::make_unique`, here and elsewhere there is a `reset(new ...)` pattern.

htuch commented on test/common/ext_authz/ext_authz_impl_test.cc:
```diff
@@ -0,0 +1,193 @@
+    response.reset(new envoy::service::auth::v2::CheckResponse());
+    ::google::rpc::Status* status = new ::google::rpc::Status();
+    status->set_code(Grpc::Status::GrpcStatus::Ok);
+    response->set_allocated_status(status);
```
Prefer the `mutable_foo` pattern to raw allocations and `set`. Sorry if I'm being a pain here, it's just a reflection of the Envoy style where we prefer to always have smart pointers to simplify reasoning about correctness. This can be relaxed in tests when really needed, but I don't think this is one of those places.

saumoh replied:
You are fine. It is more a reflection of my old-school C heritage/archaic behavior. The mutable_foo pattern and the smart pointers are much better.

htuch commented on test/common/ext_authz/ext_authz_impl_test.cc:
```diff
@@ -0,0 +1,193 @@
+  CreateCheckRequest::createTcpCheck(&net_callbacks_, request);
+}
+
+TEST_F(CreateCheckRequestTest, BasicHttp) {
```
https://23972-65214191-gh.circle-artifacts.com/0/build/envoy/generated/coverage/coverage.source_common_ext_authz_ext_authz_impl.cc.html tells me you have full coverage here, but I'm a bit surprised looking at just these tests, since they don't seem to be driving all the paths (at least directly). Could you say a few words on why we get full coverage here in this thread?

saumoh replied:
@htuch 
The underlying structures like `Ssl::Connection` context are mocks hence they are there but empty. The calls from it here thus invoks all the conditional code.
The other connditional is around `local` boolean which invokes with both `true` and `false` within the `createTcpCheck` and `createHttpCheck` it self.
The remaining conditional checks like protocol and service is invoked because those structures are getting initialized in the tests.

I've addressed the rest of the feedback as well. Thanks for the review...it has made the code much better.

htuch commented on source/common/ext_authz/ext_authz_impl.cc:
```diff
@@ -0,0 +1,191 @@
+      }
+    }
+  }
+  peer.set_principal(principal);
```
Nit: you're actually doing an additional string copy here that could be skipped if you did the `set_principal` directly in the `if` cases.

htuch commented:
@saumoh Thanks for patience during the review process, this is a nice contribution, can you rebase your other changes and get them passing tests, I'll take them one at a time? Ta!

mattklein123 commented on source/common/access_log/grpc_access_log_impl.cc:
```diff
@@ -69,7 +69,7 @@ void HttpGrpcAccessLog::responseFlagsToAccessLogResponseFlags(
     const RequestInfo::RequestInfo& request_info) {
 
-  static_assert(RequestInfo::ResponseFlag::LastFlag == 0x800,
+  static_assert(RequestInfo::ResponseFlag::LastFlag == 0x1000,
```
Seems like gRPC access logging protos should be enhanced for unauthorized? Can you add a TODO and take care of this in one of your follow ups?

saumoh replied:
yes. I'll do it in the follow up pr; thanks.
